### PR TITLE
Feature/service app gas amount

### PR DESCRIPTION
**Pull request by liayoo**
Change summary:
- Separate gas amounts, costs for service & app ops
- Separate bandwidth & state gas amounts
- Support multi-app txs

Related issue: https://github.com/ainblockchain/ain-blockchain/issues/316

### CODE REVIEW

File: common/chain-util.js [hunk 1]
```diff
@@ -324,13 +377,22 @@ class ChainUtil {
...
+    };
+    for (const [appName, appGasAmount] of Object.entries(gasAmount.app)) {
+      ChainUtil.setJsObject(gasCost, ['app', appName], appGasAmount * gasPriceAIN);
+    }
+    return gasCost;
   }
 
-  static getGasAmountCostTotalFromTxList(txList, resList) {
-    const gasAmountTotal = resList.reduce((acc, cur) => acc + ChainUtil.getTotalGasAmount(cur), 0);
+  static getServiceGasAmountCostTotalFromTxList(txList, resList) {
```
Comment (platfowner) [inline common/chain-util.js:392]: s/GasAmountCost/GasCost/?

File: common/chain-util.js [hunk 2]
```diff
@@ -324,13 +377,22 @@ class ChainUtil {
...
     if (gasPrice === undefined) gasPrice = 0; // Default gas price = 0 microain
-    return gasPrice * MICRO_AIN * ChainUtil.getTotalGasAmount(result);
+    const gasAmount = ChainUtil.getTotalGasAmount(result);
+    const gasPriceAIN = gasPrice * MICRO_AIN;
+    const gasCost = {
+      service: gasAmount.service * gasPriceAIN,
+      app: {}
+    };
+    for (const [appName, appGasAmount] of Object.entries(gasAmount.app)) {
+      ChainUtil.setJsObject(gasCost, ['app', appName], appGasAmount * gasPriceAIN);
```
Comment (platfowner) [inline common/chain-util.js:387]: minor: `ChainUtil.setJsObject(gasCost.app, [appName], appGasAmount * gasPriceAIN);`? or just
`gasCost.app[appName] = appGasAmount * gasPriceAIN;`?

File: common/chain-util.js [hunk 3]
```diff
@@ -297,21 +297,74 @@ class ChainUtil {
...
+    const gasAmount = {
+      service: 0,
+      app: {}
+    };
+    if (!result) {
+      return gasAmount;
+    }
+    const serviceBandwidth = _.get(result, 'gas.gas_amount.service.bandwidth', 0);
+    const serviceState = _.get(result, 'gas.gas_amount.service.state', 0);
+    ChainUtil.setJsObject(gasAmount, ['service'], serviceBandwidth + serviceState);
```
Comment (platfowner) [inline common/chain-util.js:341]: As discussed offline, bandwidth only would be enough for gas cost calculation by our design (the design doc needs to be updated).

File: common/chain-util.js [hunk 4]
```diff
@@ -297,21 +297,74 @@ class ChainUtil {
     return code !== 0;
   }
 
+  static isAppTx(parsedPath) {
+    return _.get(parsedPath, 0) === 'apps' || !ChainUtil.isServiceTx(parsedPath);
+  }
+
+  static isServiceTx(parsedPath) {
+    const { NATIVE_SERVICE_TYPES } = require('../common/constants');
+    return NATIVE_SERVICE_TYPES.includes(_.get(parsedPath, 0));
```
Comment (platfowner) [inline common/chain-util.js:306]: Question: Do we need opt-in based classification? Do we have any corner case that opt-out based classification (i.e. `!isAppTx()`) is not enough?

File: common/chain-util.js [hunk 5]
```diff
@@ -297,21 +297,74 @@ class ChainUtil {
     return code !== 0;
   }
 
+  static isAppTx(parsedPath) {
+    return _.get(parsedPath, 0) === 'apps' || !ChainUtil.isServiceTx(parsedPath);
```
Comment (platfowner) [inline common/chain-util.js:301]: Question: Do we need the latter case? Isn't  only`_.get(parsedPath, 0) === 'apps'` enough?

File: common/chain-util.js (hunk 4 above)
Comment (platfowner) [reply common/chain-util.js:306]: Ah, some genesis txs such as those on `/token` would be the case?

File: common/chain-util.js [hunk 6]
```diff
@@ -297,21 +297,74 @@ class ChainUtil {
...
+  static isServiceTx(parsedPath) {
+    const { NATIVE_SERVICE_TYPES } = require('../common/constants');
+    return NATIVE_SERVICE_TYPES.includes(_.get(parsedPath, 0));
+  }
+
+  static getGasAmountObj(parsedPath, bandwidth, state) {
+    const gasAmount = {};
+    if (ChainUtil.isServiceTx(parsedPath)) {
+      ChainUtil.setJsObject(gasAmount, ['service', 'bandwidth'], bandwidth);
+      ChainUtil.setJsObject(gasAmount, ['service', 'state'], state);
```
Comment (platfowner) [inline common/chain-util.js:313]: minor: `ChainUtil.setJsObject(gasAmount, ['service'], { bandwidth, state });` or
`gasAmount['service'] = { bandwidth, state };`
would be simpler? (it's up to you :) )

Comment (platfowner) [reply common/chain-util.js:313]: please ignore the previous comment. the current code just looks good,

File: common/chain-util.js (hunk 2 above)
Comment (platfowner) [reply common/chain-util.js:387]: ditto

File: db/index.js [hunk 7]
```diff
@@ -869,13 +874,14 @@ class DB {
...
       if (gasPrice > 0 && blockNumber > 0) {
         const gasCost = ChainUtil.getTotalGasCost(gasPrice, result);
-        // TODO(lia): update this to billing accounts for app transactions
-        const gasFeeCollectPath = PathUtil.getGasFeeCollectPath(auth.addr, blockNumber, tx.hash);
-        const gasFeeCollectRes = this.setValue(
-            gasFeeCollectPath, { amount: gasCost }, auth, timestamp, tx, false);
-        if (ChainUtil.isFailedTx(gasFeeCollectRes)) {
-          return ChainUtil.returnTxResult(
-              16, `Failed to collect gas fee: ${JSON.stringify(gasFeeCollectRes, null, 2)}`);
+        if (gasCost.service > 0) {
```
Comment (platfowner) [inline db/index.js:877]: What about adding the charged gas cost to the result?

File: common/chain-util.js (hunk 3 above)
Comment (platfowner) [reply common/chain-util.js:341]: So, I think this util function can provide separate values for bandwidth and state.

File: common/chain-util.js [hunk 8]
```diff
@@ -297,21 +297,74 @@ class ChainUtil {
...
+    }
+    const serviceBandwidth = _.get(result, 'gas.gas_amount.service.bandwidth', 0);
+    const serviceState = _.get(result, 'gas.gas_amount.service.state', 0);
+    ChainUtil.setJsObject(gasAmount, ['service'], serviceBandwidth + serviceState);
+    const apps = _.get(result, 'gas.gas_amount.app');
+    if (apps) {
+      for (const [appName, val] of Object.entries(apps)) {
+        const appBandwidth = _.get(val, 'bandwidth', 0);
+        const appState = _.get(val, 'state', 0);
+        ChainUtil.setJsObject(gasAmount, ['app', appName], appBandwidth + appState);
```
Comment (platfowner) [inline common/chain-util.js:347]: ditto

File: common/chain-util.js (hunk 4 above)
Comment (liayoo) [reply common/chain-util.js:306]: Right now, fees are [not collected](https://github.com/ainblockchain/ain-blockchain/blob/develop/db/index.js#L870) from the genesis txs. This is actually for tests (many of our tests allow writing at the root, e.g. /test/...). Can I make another PR for revamping the test paths & isAppTx / isServiceTx logic?

Comment (platfowner) [reply common/chain-util.js:306]: Ok. Could you leave a TODO for that?

File: common/chain-util.js [hunk 9]
```diff
@@ -297,21 +297,74 @@ class ChainUtil {
...
+      if (!appName) return;
+      ChainUtil.setJsObject(gasAmount, ['app', appName, 'bandwidth'], bandwidth);
+      ChainUtil.setJsObject(gasAmount, ['app', appName, 'state'], state);
+    }
+    return gasAmount;
+  }
+
+  static mergeGasAmounts(gasAmountObj1, gasAmountObj2) {
+    _.mergeWith(gasAmountObj1, gasAmountObj2, (a, b) => {
+      if (ChainUtil.isNumber(a) && ChainUtil.isNumber(b)) {
```
Comment (platfowner) [inline common/chain-util.js:327]: What about `return ChainUtil.numberOrZero(a) + ChainUtil.numberOrZero(b);` as undefined result is interpreted as overriding? 
https://lodash.com/docs/4.17.15#mergeWith

File: common/chain-util.js (hunk 5 above)
Comment (liayoo) [reply common/chain-util.js:302]: Added a todo (as discussed below)

File: common/chain-util.js (hunk 4 above)
Comment (liayoo) [reply common/chain-util.js:308]: Thanks, added a todo.

File: common/chain-util.js (hunk 1 above)
Comment (liayoo) [reply common/chain-util.js]: done!

File: common/chain-util.js (hunk 3 above)
Comment (liayoo) [reply common/chain-util.js]: Thanks for catching that! Updated to include only the service bandwidth to gas cost calculation.

File: db/index.js (hunk 7 above)
Comment (liayoo) [reply db/index.js]: Great idea, done!

File: common/chain-util.js (hunk 9 above)
Comment (platfowner) [reply common/chain-util.js]: As discussed offline, it seems it needs to return `undefined` for objects. So,
```
if (!ChainUtil.isDic(a)) {
  return ChainUtil.numberOrZero(a) + ChainUtil.numberOrZero(b);
}
```
seems to work.

Comment (liayoo) [reply common/chain-util.js]: Thanks for the suggestion! Changed to check isDict() and use numberOrZero().

File: common/chain-util.js (hunk 8 above)
Comment (liayoo) [reply common/chain-util.js]: Is this ditto regarding the ChainUtil.setJsObject()?

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

File: common/chain-util.js [hunk 10]
```diff
@@ -297,25 +297,62 @@ class ChainUtil {
...
+    const { NATIVE_SERVICE_TYPES } = require('../common/constants');
+    return NATIVE_SERVICE_TYPES.includes(_.get(parsedPath, 0));
+  }
+
+  static getGasAmountObj(parsedPath, bandwidth, state) {
+    const gasAmount = {};
+    if (ChainUtil.isServiceTx(parsedPath)) {
+      ChainUtil.setJsObject(gasAmount, ['service', 'bandwidth'], bandwidth);
+      ChainUtil.setJsObject(gasAmount, ['service', 'state'], state);
+    } else {
```
Comment (platfowner) [inline common/chain-util.js:316]: What about applying to only app txs (and ignoring other txs explicitly)?
```
} else if (ChainUtil.isAppTx(paredPath)) {
...
}
```

File: common/chain-util.js (hunk 8 above)
Comment (platfowner) [reply common/chain-util.js]: Ah, it was about separating bandwidth and state values. Looks like it's done already.

File: common/chain-util.js (hunk 5 above)
Comment (platfowner) [reply common/chain-util.js:302]: As discussed offline, please check if tests pass with out `!ChainUtil.isServiceTx()`.

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments (isAppTx() and getGasAmountObj()). 
Thanks!

File: common/chain-util.js (hunk 10 above)
Comment (liayoo) [reply common/chain-util.js]: 👍 done!

**Comment by liayoo** [comment]
Thanks for the thorough review! Checked that all tests are passing & now merging..

**Pull request merged**